cmd/xpumanager_sidecar/main.go

```diff
@@ -0,0 +1,355 @@
+	file, err := os.OpenFile(filePath, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
+
+	if err != nil {
+		klog.Error(err.Error())
+	}
+
+	writer := bufio.NewWriter(file)
+
+	for _, label := range labels {
+		_, _ = writer.WriteString(label + "\n")
+	}
+
+	writer.Flush()
+	file.Close()
+
+	// move tmp file to dst file
+	err = os.Rename(filePath, xms.dstFilePath)
+	if err != nil {
+		klog.Errorf("Failed to rename tmp file to dst file: %+v", err)
+
+		return
+	}
+
+	for _, label := range labels {
+		klog.V(2).Infof("%v\n", label)
+	}
+}
+
+// compareLabels returns true, if the labels at dstFilePath are equal to given labels.
+func (xms *xpuManagerSidecar) compareLabels(labels []string) bool {
+	file, err := os.OpenFile(xms.dstFilePath, os.O_RDONLY, 0644)
+	if err != nil {
+		return false
+	}
+
+	fileLabels := []string{}
+	scanner := bufio.NewScanner(file)
+
+	for scanner.Scan() {
+		fileLabels = append(fileLabels, scanner.Text())
+	}
+
+	return reflect.DeepEqual(fileLabels, labels)
+}
+
+func (xms *xpuManagerSidecar) createLabels(topologyInfos []xpuManagerTopologyMatrixCell) []string {
+	links := ""
+	separator := ""
+
+	for _, ti := range topologyInfos {
+		if ti.LaneCount < int(xms.laneCount) {
+			continue
+		}
+
+		links = links + separator + strconv.Itoa(ti.LocalDeviceID) + "." + strconv.Itoa(ti.LocalSubdeviceID) + "-" +
+			strconv.Itoa(ti.RemoteDeviceID) + "." + strconv.Itoa(ti.RemoteSubdeviceID)
+		separator = "_"
+	}
+
+	splitLinks := pluginutils.Split(links, labelMaxLength)
+
+	labels := []string{}
+
+	if len(splitLinks) == 0 {
+		return labels
+	}
+
+	labels = append(labels, xms.labelNamespace+"/"+xeLinkLabelName+"="+splitLinks[0])
+	for i := 1; i < len(splitLinks); i++ {
+		labels = append(labels, xms.labelNamespace+"/"+xeLinkLabelName+strconv.FormatInt(int64(i+1), 10)+"="+splitLinks[i])
+	}
+
+	return labels
+}
+
+// createXPUManagerSidecar creates an XPUManagerSider with an optional metrics data getter function.
+// If not given, the default metrics data getter is GetTopologyFromXPUMMetrics.
+func createXPUManagerSidecar(metricsDataGetter ...func() []byte) *xpuManagerSidecar {
+	xms := xpuManagerSidecar{}
+
+	if len(metricsDataGetter) == 0 {
+		xms.getMetricsData = xms.getMetricsDataFromXPUM
+	} else {
+		xms.getMetricsData = metricsDataGetter[0]
+	}
+
+	return &xms
+}
+
+func main() {
+	xms := createXPUManagerSidecar()
+
+	flag.Uint64Var(&xms.interval, "interval", 10, "interval for topology fetching and label writing (seconds, >= 1)")
+	flag.Uint64Var(&xms.startDelay, "startup-delay", 10, "startup delay for first topology fetching and label writing (seconds, >= 0)")
+	flag.Uint64Var(&xms.xpumPort, "xpum-port", 29999, "xpumanager port number")
+	flag.StringVar(&xms.tmpDirPath, "tmp-dir", "/etc/kubernetes/node-feature-discovery/features.d/", "location for temporary label file directory")
+	flag.StringVar(&xms.dstFilePath, "dst-file-path", "/etc/kubernetes/node-feature-discovery/features.d/xpum-sidecar-labels.txt", "label file destination")
```

Comment:
The idea was to do an atomic operation to prevent NFD from reading incomplete file. Using /tmp seems to be difficult in containers due to "invalid cross-device link" when renaming/moving the file. NFD doesn't seem to read files under directories in _features.d_, so having a temporary directory under _features.d_ maybe isn't completely out of the question?

Reply:
> BUG: NFD feature file dir should not be used as temp directory. What's the point of the whole file rename dance, if your temp file is created where NFD reads the data?

It isn't created where NFD reads the data. Read the code: `root, err := os.MkdirTemp(xms.tmpDirPath, "xpumsidecar")`

That's a subfolder, and NFD will not go there. Renaming the file is necessary for making updates atomic, not doing so would actually be a BUG, contrary to what you are claiming. The user is free to mount and use other folders, if adding a temporary folder under the default features.d causes issues, which it btw hasn't, for us.

Reply:
I know rename is needed, but I missed it using a subfolder.   IMHO it would be clearer if `tmpDirPath` would be directly the used temp dir name, e.g. `/etc/kubernetes/node-feature-discovery/features.d/xpum-sidecar-tmp`.

Reply:
Better name for the option would be `tmpDirPrefix` rather than `tmpDirPath`, because MkdirTemp() is used: https://pkg.go.dev/os#MkdirTemp

(As the destination file has fixed name, by default in same dir, I assume MkdirTemp() is used in case specified tmp dir is not in "safe" location like `/etc/kubernetes/node-feature-discovery/`.)

Reply:
I renamed the variable and tried to improve the description.